prog/vnet/rekey_nic.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+class Prog::Vnet::RekeyNic < Prog::Base
+  subject_is :nic
+
+  def start
+    nic.dst_ipsec_tunnels.each do |tunnel|
+      args = frame["payload"][tunnel.src_nic.id]
+      create_state(tunnel, args)
+    end
+
+    hop :wait_outbound_trigger
+  end
+
+  def wait_outbound_trigger
+    if SemSnap.new(nic.id).set?("trigger_outbound_update")
+      SemSnap.new(nic.id).decr("trigger_outbound_update")
+      hop :setup_outbound
+    end
+    nap 1
+  end
+
+  def setup_outbound
+    nic.src_ipsec_tunnels.each do |tunnel|
+      args = frame["payload"][tunnel.src_nic.id]
+      create_state(tunnel, args)
+      policy_update(tunnel, "out")
+    end
+
+    hop :wait_state_drop_trigger
+  end
+
+  def wait_state_drop_trigger
+    if SemSnap.new(nic.id).set?("old_state_drop")
+      SemSnap.new(nic.id).decr("old_state_drop")
+      hop :old_state_drop
+    end
+
+    nap 1
+  end
+
+  def old_state_drop
+    new_spis = frame["payload"].values.flat_map { |p| [p["spi4"], p["spi6"]] }.flatten
+    state_data = sshable_cmd("sudo ip -n #{nic.src_ipsec_tunnels.first.vm_name(nic)} xfrm state")
+
+    # Extract SPIs along with src and dst from state data
+    states = state_data.scan(/^src (\S+) dst (\S+).*?proto esp spi (0x[0-9a-f]+)/m)
+
+    # Identify which states to drop
+    states_to_drop = states.reject { |(_, _, spi)| new_spis.include?(spi) }
+
+    states_to_drop.each do |src, dst, spi|
+      sshable_cmd("sudo ip -n #{nic.src_ipsec_tunnels.first.vm_name(nic)} xfrm state delete src #{src} dst #{dst} proto esp spi #{spi}")
+    end
+
+    pop "wait_state_drop"
+  end
+
+  def sshable_cmd(cmd)
+    nic.vm.vm_host.sshable.cmd(cmd)
+  end
+
+  def create_state(tunnel, args)
```

Comment:
I admire the concision of `args`, but man, you need to be on top of things to understand how the structure of the stack get pushed allllll the way into this helper function. Maybe tests will do the job, but one of the unfortunate opportunities for confusion here is the most natural way to write the fixture here is with symbols (`{foo: "bar"}` notation) but deserialized JSON will use strings. See the fix for my messup in 5070944a8fd201227540fda972eeeef0f10b4fc7, not sure what it indicates about programming practice. It's definitely weakness on my database-lite testing methodology, to pick up more speed by getting the database out of the loop in most tests.

Reply:
Whole payload logic on strands is removed in the new pr https://github.com/ubicloud/ubicloud/pull/441. Now, it's part of the entity.